torch/nn/parameter.py

```diff
@@ -200,8 +200,7 @@ def __deepcopy__(self, memo):
 class _BufferMeta(torch._C._TensorMeta):
     # Make `isinstance(t, Buffer)` return True for custom tensor instances that have the _is_buffer flag.
     def __instancecheck__(self, instance):
-        return super().__instancecheck__(instance) or (
-            isinstance(instance, torch.Tensor) and getattr(instance, '_is_buffer', False))
+        return isinstance(instance, torch.Tensor) and getattr(instance, '_is_buffer', False)
```

Comment:
why did we remove `super().__instancecheck__(instance)` here? To check my understanding, does this mean that `isinstance(Buffer, torch.Tensor)` will now return `False`? If so, I don't think we want this to be the case

Reply:
I think removing `super().__instancecheck__(instance)` is correct, given that we're not providing the same special path for standard `torch.Tensor` as we do for `Parameter`.

There's no way to get an actual `Buffer` object; rather, the `Buffer` class just tags a tensor type to indicate it as a buffer. So we don't need the `super()` instancecheck behavior.

Reply:
The check was removed because for instance checking it seems that only the class and subclasses should be checked (covered by the torch.Tensor check) and not the parent class. The potential issue envisioned here is if the check is something like `isintance(FakeTensor, Buffer)`, then this could succeed the first check if `FakeTensor` is a subclass of `Tensor` but not of `Buffer` . Note that `__instancecheck__` doesn't take an argument for the class being checked against as that is implicitly assumed. If I understand the cpython [implementation]( https://github.com/python/cpython/blob/7c95345e4f93f4a2475418f17df5aae39dea861f/Objects/abstract.c#L2733) of `isinstance` correctly then  `isinstance(Buffer, torch.Tensor)` would check the `__instancecheck__` implementation of `torch.Tensor` if it has one or else check whether it is a subclass of `torch.Tensor` so what you pointed out should not be a problem.

Reply:
Got it, apologies for my confusion here

Reply:
Here's some code to make the issue clear for future reference if anyone looks at this, I think if someone subclasses `Parameter` they might run into this issue (results with `==` are correct and those with `!=` are incorrect).
```python
class AMeta(type):
    def __instancecheck__(self, __instance):
        return getattr(__instance, "a", False)
class A(metaclass=AMeta):
    a = True
class B(A):
    b = True
class C(A):
    c = True
def printResult(clsInst, clsCheck, expected):
    res = isinstance(clsInst(), clsCheck)
    print(
        f"isinstance({clsInst.__name__}, {clsCheck.__name__})"
        f" -> {res} { '== ' if res == expected else '!= ' } {expected}"
    )
printResult(A, A, True)
printResult(B, A, True)
printResult(C, A, True)
printResult(A, B, False)
printResult(B, B, True)
printResult(C, B, False)
printResult(A, C, False)
printResult(B, C, False)
printResult(C, C, True)
print("\n")
class BMeta(AMeta):
    def __instancecheck__(self, __instance):
        return super().__instancecheck__(__instance) or getattr(__instance, "b", False)
class B_(A, metaclass=BMeta):
    b = True
class CMeta(AMeta):
    def __instancecheck__(self, __instance):
        return getattr(__instance, "c", False)
class C_(A, metaclass=CMeta):
    c = True
printResult(A, A, True)
printResult(B, A, True)
printResult(C, A, True)
printResult(A, B_, False)
printResult(B, B_, True)
printResult(C, B_, False)
printResult(A, C_, False)
printResult(B, C_, False)
printResult(C, C_, True)
```

Result
```
isinstance(A, A) -> True ==  True
isinstance(B, A) -> True ==  True
isinstance(C, A) -> True ==  True
isinstance(A, B) -> True !=  False
isinstance(B, B) -> True ==  True
isinstance(C, B) -> True !=  False
isinstance(A, C) -> True !=  False
isinstance(B, C) -> True !=  False
isinstance(C, C) -> True ==  True

isinstance(A, A) -> True ==  True
isinstance(B, A) -> True ==  True
isinstance(C, A) -> True ==  True
isinstance(A, B_) -> True !=  False
isinstance(B, B_) -> True ==  True
isinstance(C, B_) -> True !=  False
isinstance(A, C_) -> False ==  False
isinstance(B, C_) -> False ==  False
isinstance(C, C_) -> True ==  True
```